base/rw_mutex.hpp

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include "base/assert.hpp"
+
+#include "std/mutex.hpp"
+
+namespace threads
+{
+class RWMutex
+{
+public:
+  RWMutex() : m_counter(0) {}
+  void LockRead()
+  {
+    std::lock_guard<mutex> lockWriter(m_noWritersMutex);
+    {
+      std::lock_guard<mutex> lockCounter(m_counterMutex);
+      size_t const prevCounter = m_counter;
+      m_counter++;
+      if (prevCounter == 0)
```

Comment:
Imagine that two readers calls LockRead, and it's possible to determine that, say, the first reader calls m_noReadersMutex.lock(). Then, the first reader calls UnlockRead, but m_noReadersMutex is still locked by the first thread (because the first reader called m_noReadersMutex.lock). Then, the second reader calls UnlockRead, so, the second thread calls m_noReadersMutex.unlock(). In accordance with cpp documentation, this is UB (http://en.cppreference.com/w/cpp/thread/mutex/unlock).